### DISCUSSION TITLE
Terrain

### DISCUSSION THREAD

**Discussion by HogJonny-AMZN** [o3de/o3de#1974] [General]
We have some design and did some early prototyping for terrain content tools and workflows.  Essentially it is a design for componentized terrain data.  We can use this discussion thread to discuss this topic more with the o3de community.

**Comment by HogJonny-AMZN** [comment]
Here are some of the early ideas we had about modernizing the LY terrain workflows.
[lmbr-TerrainFeatureNotes-080721-1637-228.pdf](https://github.com/o3de/o3de/files/6786024/lmbr-TerrainFeatureNotes-080721-1637-228.pdf)

**Comment by HogJonny-AMZN** [comment]
Componentized Terrain looks something like this, where everything is asset backed and works with prefabs, segmented worlds, multiple-users editing, etc.
![Lumberyard Terrain Feature Spec](https://user-images.githubusercontent.com/23222931/124960209-b313ff80-dfe1-11eb-90dc-ab8d0c69dec8.jpg)

**Reply by robinm235** [reply]
seems like a good approach to me, allows for a lot of flexibilty especially with multiple people working on a world

**Reply by WashedUpStudio** [reply]
yes I agree this would be a great addition to the terrain system when it is added.

**Comment by HogJonny-AMZN** [comment]
Basically with the prototype, you could use a shape as a stamp, or a image gradient+box, etc.
The idea is something like a house as a prefab, can have a shape with falloff that modifies the terrain (to flatten it).
Also this whole concept supports the idea of everything being streamable.

**Comment by robinm235** [comment]
the "Terrain-Stamps" from the PDF sound like an extremely useful tool for terrain editing.

would they support Alphas/Heightmaps to place like a detailled mountain into the terrain ?

**Reply by HogJonny-AMZN** [reply]
Yes, you can use a shape to modify the terrain.
You could have a polygon prism for instance around a house in a prefab, and generate falloff, and use this to modify the terrain.
You can have a image (as a gradient) bound to a box and composite to modify the terrain.
You could 'brush' data into a component and composite that data.
These things could be part of a prefab, copies of the prefab carry these edits.

**Reply by mbalfour-amzn** [reply]
Just to provide a bit more of a tangible view into what this could look like, here's a screenshot from some early prototyping I did:  
![image](https://user-images.githubusercontent.com/82224783/125860021-c41e4310-b5f6-4163-b2f6-609c16e51f10.png)

The idea is that you would have entities that define regions through Shape components (Boxes, Spheres, Polygon Prisms, etc).  Each of those regions would have a height provider, which is basically the Gradient components (Image Gradient, FastNoise, etc).    The example in the picture shows a couple of FastNoise regions and one Sphere Gradient region.

@jeremyong actually from this picture, you can hopefully see how it would be possible to provide cube tiles, honeycombs, etc.  

So, for terrain "stamps", what that could/would look like is using the Gradient components to build up a compositing network where you could take an Image Gradient (heightmap) as a base, use various types of gradient modifiers to blend with shape falloffs, stamps, inversions, etc, and then feed the result outward as the height provider.

There might be a way to do it with just overlapping regions as well, but ideally we won't tackle that in v1, because I have a feeling it will just lead down a path of almost identical chained blend operations as what Gradients already provide.

**Comment by jeremyong** [comment]
Three topics for consideration:

1. Having an offset tile pattern (honeycomb) for streaming purposes. It's easier to consider which tiles should be made resident more conservatively if the tiles aren't meeting at corners in a square tiling pattern.
2. Using mesh representations for terrain (and taking a more kitbashing approach to authorship). Requires a good hlod, culling, and rasterization solution but avoids heightmap restrictions. (This in addition to any heightmap triangulation)
3. Integration with ray tracing

**Reply by HogJonny-AMZN** [reply]
1. This workflow is heavily influenced by the numerous terrain editors that largely segment worlds into grids.
We work backwards from that authoring experience, and it makes our approach highly compatible with basically any image based authoring tool (houdini, world machine, instan terra, etc, etc.) Also we hope that people could easily port existing world data in.  You can also literally paint and use images from photoshop and other tools.  In many ways, bounding boxes makes sense to artists and this type of data.
I understand the streaming comment, we might want to considering doing something like that as a structure to inform data that needs to be streamed, however I think I would find it overly difficult and restrictive to 'have' to author that way.
2. many teams have gone the mesh based world approach.  we really want the authors to choose, they could blend the approaches (which is pretty common)
3. understood, we need to adapt the creative first needs to the tech and consider perf, etc.

**Reply by mbalfour-amzn** [reply]
Ideally, the design won't _dictate_ the use of a grid of tiles, but as @HogJonny-AMZN says the grid is the generally-expected use case.  The approach I'm considering has a little bit of separation between data _providers_ and the data _consumer_, so even if the underlying system is querying the data and using it in an aligned-box fashion, the providers themselves wouldn't have to be arranged that way.

**Reply by mbalfour-amzn** [reply]
Ray tracing is a super-interesting question for this.  I don't think we're going to want to tackle ray tracing as a part of the initial rendering implementation, but I'd like to structure and separate the various parts of the system enough that the rendering side could be swapped out for it at some point without impacting the other parts.  There's a semi-recent paper that explores real-time terrain raytracing through a single fullscreen quad approach ( https://fenix.tecnico.ulisboa.pt/downloadFile/281870113703784/ExtendedAbstract-Final-73465.pdf ).  Not sure how nicely that would play with other things in the Atom scene, but there's definitely promise there.

**Reply by Infurael** [reply]
The edit time shouldn't dictate the runtime and vice versa. Using tiles in the editor shouldn't mean that the runtime data uses tiles, but can use other structures. The use of tiles in the editor should be aimed at supporting a particular artist workflow. Behind the scenes the terrain system could take terrain data (no matter how it's split up) and composite the runtime data that fits the output best.
In general tying the edit time and runtime together means compromises need to be made leading to either a suboptimal user workflow or suboptimal runtime performance. A similar philosophy has been applied to Prefabs and it already resulted in Networking having greater flexibility. Consider for instance the use by simulation customers that may just have a huge fbx file (assuming terrain uses information from such files) that needs to be split into tiles for the runtime, or companies that like to use the tile approach during authoring while having a different approach for the runtime.

Using tiles for the runtime is not exclusive to terrain, but has been a general project for Streaming. The idea is to use the new Prefab processing stack and split prefabs into tiles behind the scenes then have a second pass combine all individual prefab tiles for the same tile together into a single spawnable tile. If the terrain is going to support tiles for the runtime those two efforts should be aligned so there are not two competing tiling solutions.

Based on experience working with tiles at edit time, I know it has problems scaling to larger teams. Tiles create quite artificial borders which often don't line with natural looking terrain. As such artists and designers are forced to check out tiles they're only partially using which locks others out of that region. Adding layers helps mitigate some of the problems by allowing different disciplines to work on the same area but that also comes with problems such as misaligned layers.
It's not a bad place to start though, but since this is a new terrain system there's an opportunity to create a different workflow and address some of these problems. Then again, keeping with the modular nature and split between edit/runtime in mind as well, the tiling approach could be a start with alternative workflows offered in the future.

**Reply by mbalfour-amzn** [reply]
@AMZN-koppersr I wholeheartedly agree with the overall sentiment that edit and runtime shouldn't be coupled if it leads to compromises.  That said, I _am_ proposing an initial coupling between edit and runtime representations, though it's a bit more nuanced than "tiled", and I don't think it's causing any specific compromises.  I also believe they could be decoupled later on a case-by-case basis if needs dictate.  Let me know what you think of the following...

The data providers themselves are entities with arbitrary regions.  They can be laid out as tiles, scattered regions, overlapping regions, etc.  I would initially at least limit the region shapes to AABBs for simplicity, but even that isn't really a _hard_ requirement, it's just avoiding some initial math complexity in the implementation.  The data _inside_ the regions can be any shape, so it shouldn't be a very burdensome requirement.

These would be the same regions that are used at runtime, so in that sense, they're tied together.  

However, there's a bit of decoupling that occurs both _above_ and _below_ in this design.  The data provider entities _themselves_ can get divided up behind the scenes into streaming tiles the way you describe.  I'm not sure what the streaming plan is for entities that have bounds that span across multiple streaming tiles, but ideally whatever that plan is, it would apply here as well.  Depending on that design, it may or may not be invisible to authors.

Within each data provider, streaming can look different depending on the data.  Procedural providers are streaming by nature, because they're creating data on-demand, with little to nothing stored.  Image-based data providers might use streaming mip levels and/or virtual textures via texture streaming.  Real-world GIS data might be getting streamed from the cloud asynchronously.  The concept of "streaming" at this level can be extremely different and specific depending on the type of data.

On the rendering side, terrain would likely get rendered as patches of various LODs, but those patches will be composited and rendered on-demand based on visibility.  Not exactly _streaming_ in the way you're talking about, but it's still a similar concept of data existing and ceasing to exist in memory over time.  The compositing is based on the existence and data fidelity of the various data providers, so it's indirectly connected to the streaming concepts above, but not directly connected.

To one of your last points, about tiles (or other region-based representations) causing artificial barriers and workflow issues, the compositing and separated data channel support really should help address this.  Within a region, it would trivially be possible to have people separately working on height data, color data, and surface data.  If a team needs a more complex collaboration workflow, they can use compositing to feed multiple layers of height data together, and have individuals work on each layer.  If we're talking 1 km^2 regions, that should be fine, but if the goal is to have multiple people working on the same square meter, that's still _technically_ possible, but questionable. :)  They can also start with a grid of say 1 km^2, and then just overlay any number of small 5-10m regions in the localized areas they're working on.

**Comment by jeremyong** [comment]
Question: How will the channels of the terrain integrate with the material system? I imagine artists may want to author custom materials/shaders that write into the channels of the terrain virtual textures.

**Reply by HogJonny-AMZN** [reply]
The prototype used splatmap inputs in a component (with n# configurable layers), but because it was the legacy terrain rendering these inputs informed terrain detail material placement/masking/splat layers.  each layer referenced a material, these detail materials and composited/blended with the macro inputs. with Atom we have more flexibility, this design isn't set in stone, it's just what we know works and is well understood - and we felt it could be virtualized.  Also there is edit time flexibility and the need to be dynamic, but the data could be baked down for runtime (so condition it however we need) Or have portions baked to runtime data but allow some dynamic edits to continue to occur.

**Comment by WashedUpStudio** [comment]
For terrain manipulation making holes in the terrain is a must, so you can add things like cave meshes, etc. but making it possible to do underpasses for large cliffs, terrain built caves, etc. would be an amazing feature that I do not see in many terrain editors for engines. (at least not out of the box)

**Reply by HogJonny-AMZN** [reply]
Yes - we always have intended to continue to include holes. But there are also other ways to tackle this, like replacing a patch with some kind of voxelized structure.  But organic painting tools would be awesome (in my mind something like vdb+sdf+meshing, either at runtime or baked.) With soft blending you can seamlessly transition meshes into the existing world (like terrain).

**Comment by WashedUpStudio** [comment]
Robin posted "unreal has a quite interesting new approach there called virtual heightfields, its basically a much better way of tessellating a terrain to give it more geometry detail, follows the used materials heightmap". It looks like a great feature as well.
![image](https://user-images.githubusercontent.com/66225183/125003447-17e44f80-e00c-11eb-8aaf-2f03626c65cc.png)

**Reply by mbalfour-amzn** [reply]
This looks a bit like a cross between https://www.gdcvault.com/play/1026349/Advanced-Graphics-Techniques-Tutorial-Four and http://twvideo01.ubm-us.net/o1/vault/gdc2015/presentations/Chen_Ka_AdaptiveVirtualTexture.pdf .  Basically, using virtual textures both for heightmap tiles as well as the composited terrain textures, but then also layering per-material height into the heightmap tiles as well for additional tesselation.  This is pretty much the path I've been thinking about for O3DE terrain.  One caveat - in the Unreal version, the material height compositing is happening render-side, which means that it isn't readily available to physics or generalized game queries, it's just a render-side effect.  Not bad for tiny amounts of geometry, but not good for something like the giant "cobblestone" mounds above.  I'd like to make sure that composited data exists CPU-side so that it's at least _possible_ to get physics to line up with it.  Whether or not it's a _good_ idea to feed physics that level of tessellation is another story, but to be fair it's also probably not a good idea to make your materials affect heights as much as you see in that screenshot. :)

**Comment by robinm235** [comment]
Having some Sort of voxel terrain Option would be awesome, especially if it could be combined with cheaper regular terrain in areas where No voxels are needed.

For example, if we could create something like "VoxelPatches" to sculpt maybe a cave or cliff into the terrain and then bake that for runtime to make it cheaper and blend it in with the surrounding image based terrain, that would be pretty much the dream for level design

**Comment by mbalfour-amzn** [comment]
Just to take a step back here for a sec, I'd like to define what is and isn't a "terrain system".  To me, a terrain system covers the following:
* Authoring (height, color, surface types)
* Rendering (including culling, LOD blending, shadowing, etc)
* Simulation (physics, general height / color queries)

It's _related_ to but slightly separated from vegetation, decals, roads, editor placement tools, and streaming.

I'll put this out there for discussion before going into any more thoughts on what we could do to address these areas...

**Reply by bruluk5w** [reply]
Well, regarding decals and roads and such, I'd argue it depends on whether it could be baked into the terrain for rendering (whether to actually bake could be a choice based on resulting data size vs. runtime peformance impact and flexibility of non-baked items).
Maybe vegetation could be a different component that uses an api exposed by the terrain system to handle changes on any of the layers that the terrain system uses. These would most basically be changes in the heightmap, but it could be any data layer that just happens to be assigned by default or by the user to the vegetation system. This would also allow other assets to use terrain updates, such as unique meshes that reposition themselves when the terrain height changes (e.g. lamps, curbstones and signposts along steets). For steets there could even be a different system (or even gem?) that provides tools for editing steets and that writes data to an assigned layer in the terrain system.
Despite technically maybe being different systems, from an artist's point of view I guess it would be nice to have these things avialable within the same editor interface.

**Reply by mbalfour-amzn** [reply]
Good point, let me be more specific.  I see decals and roads as separately-authored items that interact with terrain, because you often want them to "snap to terrain", be blended with terrain from a rendering perspective, and potentially affect the physics surfaces broadcast at those points.  These systems need to know about terrain and interact with the terrain, but ideally the terrain shouldn't need to directly know about them.  It just needs to expose enough APIs to allow for those interactions.

Vegetation is a very different story.  We've got a dynamic vegetation system that uses a series of Gradient and Surface Data components to determine where and how to place vegetation.  The idea for terrain, that I'll post here in a second, is that terrain would _also_ use those components.  So it's not that vegetation knows about terrain exactly, it's that we have a way to optionally provide the same source data to both systems to interpret in different ways.  We named the "Landscape Canvas" tool the way we did because even though it only affects dynamic vegetation right now, our intent is that over time it will be extended to support terrain and other related layout tools as well.

And I 100% agree these systems should all be presented in the same editor with unified workflows. :)

**Comment by bruluk5w** [comment]
I was reading the pdf was wondering where the quadtree was 😅.  Concretely I'm thinking that the proposed approach is very versatile in what features it may support, but I think it could maybe be extended to employ a hierarchical subdivision on some level. Else I think the solution could be limited in the maximum size of the terrain. GPU tessellation is limited and just decreasing the terrain resolution to have a larger area to work with is maybe not an option if you still want to paint a road material with tessellated and displaced cobblestone on the road.
So I just wanted to kick of thinking big. Maybe it's not what the engine will aim for, but if there is a chance to build something new, why not thinking in size of planets or even infinite terrain with a procedural data source? What are your thoughts? Could the terrain system handle online map tiles of the earth?

**Reply by mbalfour-amzn** [reply]
Hah, fantastic!  The proposal that I'm writing up is thinking _big_, but with some caveats.  O3de doesn't support spherical coordinates, and it currently has the standard precision problems that come from floating-point positions beyond 16 km.  

I'd like to take "planetary support" and spherical coordinates out of the mix, because that's a problem that requires support far beyond terrain, and it's a system choice that would be more difficult for non-planetary-sized applications (the more common use case) to work with.  There's absolutely a need for a system like this, but I think it would need to be a separately-authored system specifically aimed for that problem set.

So, coming back to "flat planet" solutions where it can stream online map tiles, or possibly even procedural infinite terrain, _that_ should be doable with what I'm proposing.  It would require a developer to write the higher-level system for identifying and streaming the tiles, or writing the procedural algorithms, and more importantly, it would need to handle the rebasing / camera problems to avoid the 16 km precision limits, but the terrain system itself should handle it.

(As a side note, if you or someone else wanted to tackle any of those problems in o3de, I've got a lot of thoughts on how to approach them :) )

I was just in the middle of typing up my system thoughts as a new post, so let me finish that up then come back here to describe some of the details relative to what you're asking about.

**Reply by bruluk5w** [reply]
Yes, I also agree that planets are maybe off the table (bent tiles, singularities etc. make this difficult, you named it).

**Reply by mbalfour-amzn** [reply]
So, coming back to this, where you're asking about quadtrees and subdivision, max size limitations, and tessellated cobblestones...

Quadtrees / subdivision - this would come into play at a few levels.  The coarsest-level subdivision is the separate entities for data providers.  They provide a level of manual control for subdivision.  This helps with simultaneous authoring in different parts of the world, high-level streaming, etc.  Physics has its own internal structures for heightmap storage, so we'll skip discussing that one.  So that brings us to rendering.  The terrain rendering code would likely use a quadtree (or an octree) for LOD management of different patches of the terrain.  Inside the rendering, there will also likely be some form of virtual texturing, which could leverage that same quadtree or potentially a different hierarchical structure (along with the obvious indirect texture itself).  Atom doesn't have generalized virtual texture support yet, so there might need to be a few development passes on this.

Max size limitations - the goal is for the system to not have any hard-coded size limitations.  The limitations should come from user-set parameters (grid size, virtual texture size, etc), available disk space for assets, available RAM, available CPU, coordinate system precision, etc.  Basically things that should be able to fluctuate by use case and by time, not things that suddenly require massive system rewrites.

Tessellated cobblestones - I'm calling this one out separately because it has some implications.  There's tessellation as a pure render-side artifact, and there's tessellation where it can be physicalized and queried by arbitrary game systems.  i.e. Is it "terrain system" tessellation, or "render" tessellation.  The goal for the terrain system is to potentially allow for physicalized tessellation of those cobblestones.  It could either be handled through height data providers and a high enough grid resolution that the physics heightmap can have cobblestone-level differences, or it could be through material data and a custom physics solution, or probably a few more combinations as well.

**Reply by kberg0** [reply]
Honestly we probably should be forward looking towards solutions that scale to planet size.. otherwise we'll likely find ourselves right back here in 6 months to a couple years rewriting terrain yet again. Even if we don't solve for all the issues you mention up-front, having something that can handle moving from ground level to flight sim and LOD'ing appropriately seems quite reasonable.

**Reply by mbalfour-amzn** [reply]
The goal I'm working towards here is to deliver an initial terrain system for our o3de customers, where right now they don't have one at all.  The more we increase the scope, the greater the funding and/or community involvement needed, and the longer it will be before our customers have a terrain solution.

LODing from ground to flight _seems_ quite reasonable, but in my experience, mostly-ground-based games have an extremely different terrain system than flight-based games.  Ground-based games are typically focused on mm-resolution terrain fidelity and often don't need more than 16 km^2 of terrain.  Flight-based games are focused on speed of streaming and view distance, often need hundreds or thousands of km^2 of terrain, but generally don't need fidelities below meter resolution.  

I'm admittedly biasing more towards ground-based games here, as they're the _more_ common of the two needs, and until the engine supports high-precision coordinates and generalized streaming, I believe there are some very real barriers that would make it difficult to even build and test a terrain system optimized for flight.

**Reply by bruluk5w** [reply]
I would agree that terrain should at least scale up to the point where fast movement high above the ground should still look as expected by an artist. When we say it has to scale for flight sims, then i think we have to make sure we are speaking of the same thing: flying planes in Battlefield looks cool but doesn't need to account for map projections and all associated problems. MSFS has to and Star Citizen takes it to another level. In the second one we could see how resource intensive it can get once you do not have a flat-earth-assumption anymore. For the first one i can tell you from experience that just one planet is a very involved process. They did an amazing job with their world streaming but it comes with performance implications that any non flight-sim developer would probably not take. Also it basically alters the whole content creation pipeline, because you place your cities, houses, cars etc. at something like lat/long(/altitude) with reference to a planet instead of x/y/z which has to be converted to cartesian during runtime. Additionally normal square tiles in a mercator projection would produce ugly distortions in high latitudes and any tessellation of a sphere that doesn't have such problems will include non-rectangular tiles (apart from the cube), which do not fit well with how the gpu works. But once we go to planet size we basically have to have a hierarchical organisation of variably sized tiles also on the asset level, which i am out of ideas how this would be efficiently done with non rectangular shapes. Hm, idk if it makes sense what i'm saying, but basically supporting planet sized terrains that are flat and just wrap around on the borders should be doable. But introducing spherical coordinates in the engine doesn't make sense in my opinion. But maybe my horizon is limited even though i said we should think big. If there is a system that could do this without big compromises, then I argue that it's design would take unproportional amounts of time.

**Comment by mbalfour-amzn** [comment]
OK, I think my definition of a terrain system has been mostly non-controversial, so here's the more specific thoughts around what I'd like to put in the RFC.

For authoring, the idea is to follow a similar paradigm to the Dynamic Vegetation system.  Gradient components would be used to separately define height data and surface data weight maps.  New "spawner"-style components would be used to map gradients to areas, and in the case of color or surface data, which materials they apply to.  Note, this is all just for _defining_ data and how it maps (and indirectly how it can be loaded/unloaded via dynamic spawnables); this is separate from rendering and physicalization.

Physicalization would have spawner component regions to spawn physics heightfields using the height and surface data from the entities listed above.  Rendering would have spawner component regions to render the height, color, and surface data from those entities.  By making these spawners separate, it adds a bit of complexity, but it also adds flexibility because it allows for areas of terrain to be rendered without physicalization and vice versa.

Finally, the back-end terrain system would provide a consistent world grid and manage the data necessary to interconnect with rendering and physics.  It would also provide the API for generalized editor and game system queries.

Putting this all together, in the super-simple case of a single non-streamed terrain, you'd just have a single entity with a world box, a set of terrain, physics, and rendering spawners, some child gradients that take in heightmap images and splat maps, and some physics and rendering materials.  

In the procedural case, instead of heightmap images via Image Gradients, you'd just use something like the FastNoise Gradient to procedurally generate the noise function of your choice.  Or possibly a new custom Gradient component doing a specialized form of generation.

In the case of a more complicated streamed terrain, there would be sets of entities.  For simple streaming, you'd make the entities be non-overlapping boxes, maybe 1km^2, to whatever grid size you want.  These would all be saved as dynamic spawnables, and a simple script or something could load in the N closest grid entity spawnables around the camera as the camera moves around.  In the procedural case, you'd just keep spawning the new instances of the same spawnable over and over N times in the right grid location, and the proceduralism will take care of producing new unique data.

By separating out height, color, and surface providers, this also allows for more complicated things like the "terrain stamps" from the initial doc at the top of this.  You could create a dynamic spawnable with an entity that contains a cottage, and a height gradient provider that masks out some level ground around the cottage with some falloff.  This could get loaded at artist-defined or random locations, and at runtime modify the terrain to smooth itself out around the cottage without necessarily affecting color or surface data.  Or the spawnable could contain those as well, to ensure that the cottage always has a manicured lawn.  etc.

One other point that this all implies - the data can be authored at different resolutions for height vs color vs surface, at a per-entity level.  So you could have playable areas use a much higher density of height data than non-playable areas, for example.  Though for simplicity, I think the back-end system should manage a "world grid" at a consistent resolution independent of the input data resolution.  This just makes things like seams, blending, etc a non-issue.

**Reply by mbalfour-amzn** [reply]
Oh - I should also explicitly call out that in the initial implementation, I _am_ proposing a heightmap-based approach at the core, not a meshified approach, TINs, voxelization, or raytracing.  Frankly, this is because it's simpler to author, manipulate, and write, meets a majority of use cases, and plays nice with things like PhysX.  However, by separating apart the various types of input data, and keeping _those_ separate from the output systems, and by making a 3D-based querying API, it should be possible to extend the system design over time to use alternate data source as inputs, and alternate renderers and "physicalizers" on the other end.

**Reply by bruluk5w** [reply]
I am not sure if I understand some parts. With physicalization you mean turning the data sources into a mesh representation? I've been working in games for about 3 years but terminology is sometimes not ubiquitous or the same ;)
Also does the proposed approach mean that all loaded terrain has the same resolution? Because that would mean that if someone authors a big mountain in the far distance, it would only appear if the player approaches near enough to it.

**Reply by mbalfour-amzn** [reply]
"Physicalization" - I just meant "can be used with physics".  I'm thinking we would specifically use a physics heightmap, but it could also mean a physics mesh, a physics convex hull, etc.  I wasn't trying to name a specific implementation, just describe the usage.

Does all loaded terrain have the same resolution?  Not exactly.  This is a complicated question, I'll try to "walk the path" of the data in my answer.  I'm also just going to focus on height data, since that's the easiest to describe, and similar concepts will apply to the other types of data.  To start with, height data can either be procedurally generated or artist-created.  From a Gradient Component point of view, here's what those look like:
![image](https://user-images.githubusercontent.com/82224783/126382024-2d579f11-d9ce-4e33-b0dd-55d07febedee.png)
![image](https://user-images.githubusercontent.com/82224783/126382069-23ca4865-335e-4fc3-a3ef-aeffc104d52b.png)

Procedural data can have "infinite" input resolution.  Artist-created data has an input resolution dictated by the resolution of the input image.  So at the data input level, every gradient loaded can have a different resolution.  Some parts of the terrain could be at 0.25 m, some could be at 1m, some could be at 32 m, for example.

_However_, the part of the system that transforms this data into physics or rendered meshes needs to pick a resolution for that transformation.  Let's also say that on the back-end, all terrain is getting divided into 32m x 32m patches, and each patch can be a separate LOD.  And for simplicity, let's say physics and rendering use the same LOD schemes.  So, it could be that the nearest 1 km x 1 km of patches will get turned into heightmaps with 0.5 m resolution.  The next 1 km of patches is 1 m resolution, the next 1 km out is 2 m resolution, etc.  

What this would mean is that the input data is getting queried and quantized at those transformation resolutions, regardless of the input resolution.  So that 0.25 m data will be a waste, because it will always be used at 0.5 m or greater, and the 32 m data when it's really close will be _tessellated_ to 0.5 m, even though it will still have 32 m granularity.  However, if you changed your tessellation to 0.25 m, you would suddenly see more detail in that 0.25 m area.

Where things could get a little more complex is if we support mipmapping on the Image Gradients (we currently don't).  If we add that, we could have a level of dynamic LODs on the _input_ data as well, based on the mipmaps streaming in and out.  I think that could be a cool additional level of control, but let's put that back off to the side for the moment.  I just wanted to call it out as another possibility here.

For that big mountain in the distance, the visibility of that first depends on whether or not it's loaded at _all_.  Once we've got an entity loaded that covers that region with data, then the visibility comes down to how its rendered.  Assuming the rendering has an LOD system and a far enough draw distance, the big mountain would still appear, just at a lower resolution.

At some level though, yes, there will be a mountain far enough away that you simply can't see it, regardless of resolution and LOD.  In curved-world scenarios, that would _also_ occur because it's below the horizon line, but here it's because there will ultimately be a limit to how much we can load and draw at any one time.  IMO, the fancy thing to do would be to semi-simulate horizons and have the distant items "grow" in instead of popping or fading in.  I believe something like that could be done as polish work on a system like I'm describing.

**Reply by bruluk5w** [reply]
Thank you for the details, now things make a lot more sense and from a general point if view (since I'm not an expert on the engine) i think it sounds like a promising starting point.

**Comment by ppinfel** [comment]
for my terrain purposes i only need a specialized mesh component that only takes a heightmap and gives u a highly performant static mesh, dunno if that is possible though

**Reply by mbalfour-amzn** [reply]
:) My proposal is definitely more complex than that, but at the end of the day that should be possible with what I'm proposing with a 1-2 entities and a handful of components.

**Comment by kberg0** [comment]
A couple things that jump to mind on a quick skim, note that I haven't had a chance to really deep dive the proposal, so my apologies if these points have already been considered.

Increasingly we see dynamic interactions and voxelization (non-planar, non-heightmap based solutions), since heightmap based solutions have inherent slope limits based on your terrain resolution, lack of overhangs, lack of caves, etc.. that might require a hybrid approach where we use height-map based terrain within most tiles but allow for a height-map to be swapped out for something more voxel friendly. This is a non-trivial market segment for game engines as nearly all survival crafting genres depend on this to some extent.

I'm cringing over the heavy usage of the entity component system, anticipating performance issues. The vegetation system as an example is never going to hold up to production scales given our current implementation of entities and components, and I anticipate vegetation is going to require a pretty hefty refactor before any customer can ship a game using it. What are we doing to mitigate this, since we know it's a really serious concern up-front.

Somewhat related, I'm concerned about the scaling of this solution from ground level to flying. Even simple shooters nowadays typically mix some combination of troops on the ground with carriers or ships or fighters in the sky. If individual tiles are entities, a solution that performs well on the ground could quickly fall down once in the air and suddenly millions of terrain tiles are within the view frustum. It seems like we need a better form of CLOD here.

**Reply by mbalfour-amzn** [reply]
Dynamic interactions are accounted for in the design, data provider changes would trigger recompositing of the data.  Voxelization has been _considered_ but isn't being _proposed_ as a part of the initial implementation, because it brings additional complexities such as potentially needing changes on the physics side of the engine.  There also just aren't as many authoring tools for producing voxelized data.  For 3D deformation, there are also other options that are appropriate in various scenarios, such as deformation maps / image-based warping, TINs, etc.  The goal is to provide a framework and APIs that can be extended into 3D over time, but not to _immediately_ tackle that.

Entity component system - I totally get your concern here, though it would be nice to put some hard numbers to this, and to split the concerns from vegetation and terrain.  Right now, the vegetation system performance is completely bottlenecked by a few known performance hotspots in Atom, not by the entity component system.  Based on how it performed with the previous renderer, the _expected_ hotspots come in with extremely complex layered data designs.  We've got a design and prototype for baking out a number of layers at edit time, which greatly reduces the runtime complexity and quickly turns back into a rendering bottleneck.

For terrain, if we're talking a few entities per square km to describe the terrain, can you elaborate on what your specific concern is?  If we can't support that level of entities with our current system, we're going to have an exceedingly hard time _populating_ the terrain. :(  

To the final point, I definitely share your concern with how this scales to flying.  Flight will exercise this system with completely different data sets, use cases, etc.  Data decompression, streaming speeds, compositing speeds, and overall numerical precision are all going to be challenges.  FWIW, the "millions of tiles" case should be pretty rare though.  If you make your tiles 1 km^2, a million tiles represents 1000 km^2, which means you're looking down on an area representing 1/4 of the USA.  I think 65536 tiles (256 km^2) or 256 tiles (16 km^2) would be more the order of magnitude we're talking about.

**Reply by HogJonny-AMZN** [reply]
In the GDC2019 Dynamic Vegetation demo, most of the 'world data' for biome placement was represented by a fixed number of Image Gradient inputs, basically if you want to place 5 biomes, the placement data can be fully represented by 5 entities: each a box + image gradient.  These actually used a 'reference shape' to share the world box.  Each biome itself can be as simple as a single entity with a 'layer spawner'. But the complexity of the number of entities driving inputs to the composition is arbitrary and up to the user.  The placement system only needs to query positional data.

The performance nature of the vegetation models and rendering themselves, is an entirely separate concern and was the vegetation system itself (not the dynamic vegetation systems.)  Managing large numbers of objects, batching (or merged mesh), etc., that is bottlenecked currently I believe by defining the things to be placed as entity/prefabs ... so if we place 100k plants we have 100k prefabs.  There is versatility to being able to procedurally place a prefab of arbitrary complexity (for instance, a boulder mesh + collision component).

When it comes to drawing 100k copies of the same mesh, we need better ways of doing that... but the actually number of entities needed to **describe the world** is very manageable already and relatively low in comparison to the complexity of the rendered world itself.

**Comment by hollsteinm** [comment]
Not sure if this is relevant to a terrain editor in this scope but I figured I would drop this thought here. Would there be support for more than just a flat landscape editor, i.e. perhaps building a terrain that can be applied to a sphere or some shape to make building multi-world terrains within a universe a first class citizen? Brining this up is because some game worlds are more than just one landscape, or a single landmass and enabling the authoring of such worlds would attract those looking to build larger simulations in that vein.

**Reply by hollsteinm** [reply]
And just to be clear, the shape is more an abstract concept for traversal and streaming purposes. I think the concept of a "complete" world is what is more important to capture. I.E. building in a feature that enables "wrapping" between landmasses but still maintain for simplicity and industry standards the flat grid like terrain.

**Reply by mbalfour-amzn** [reply]
It's an absolutely relevant question, thanks!  The intent here is to develop more of a "flat Earth" / "unwrapped Earth" style terrain system, not one that's wrapped around a planetary shape.  As @bruluk5w covered in their discussions above, there's definitely a use case and need for planetary-style landscapes, but it can be significantly more difficult to work with as a system.  "Wrapping" a landmass around a shape requires different coordinate system support, non-uniform data sets to avoid warping, completely different physics support, etc.  Supporting this type of arbitrary wrapping is definitely the more generalized solution, since you could wrap a plane to produce results equivalent to a flat Earth, but since it's been generalized to that, you're still taking the performance and complexity hits of not _assuming_ it's wrapped on a plane.

For an initial terrain offering, I'd like to stay away from imposing those additional difficulties on the majority of customers that _don't_ need that level of complexity.

If somebody _else_ who's working on a planetary-style game or simulation wants to propose a terrain system for that use case, I will be more than happy to work with them to align the system designs to make it as easy as possible to share technology and data between the two.

**Comment by mbalfour-amzn** [comment]
It seems like this discussion area has mostly reached the limits of what we can talk about without a concrete design, so I've posted an RFC here:  https://github.com/o3de/sig-content/issues/4 .  Let the commenting commence! :)
